chia/full_node/full_node.py

```diff
@@ -358,15 +358,21 @@ async def manage(self) -> AsyncIterator[None]:
                 if self._transaction_queue_task is not None:
                     self._transaction_queue_task.cancel()
                 cancel_task_safe(task=self.wallet_sync_task, log=self.log)
-                cancel_task_safe(task=self._sync_task, log=self.log)
+                for one_sync_task in self._sync_task_list:
+                    if not one_sync_task.done():
+                        cancel_task_safe(task=one_sync_task, log=self.log)
 
                 for task_id, task in list(self.full_node_store.tx_fetch_tasks.items()):
                     cancel_task_safe(task, self.log)
                 if self._init_weight_proof is not None:
                     await asyncio.wait([self._init_weight_proof])
-                if self._sync_task is not None:
-                    with contextlib.suppress(asyncio.CancelledError):
-                        await self._sync_task
+                for one_sync_task in self._sync_task_list:
+                    if one_sync_task.done():
+                        self.log.info(f"Long sync task {one_sync_task.get_name()} already done")
+                    else:
+                        with contextlib.suppress(asyncio.CancelledError):
+                            self.log.info(f"Awaiting long sync task {one_sync_task.get_name()} to finish")
+                        await one_sync_task
```

Comment:
the indentation level of this line appears to be wrong.
```suggestion
                            await one_sync_task
```

Reply:
i shortened the comments and ya i fixed the indent. oooof

Reply:
So this PR is trying not to change the existing behavior of the sync at all, because it works. It just tries to hold on to references to tasks at all costs. The issue is that multiple tasks can arrive at the task creation, thus overwriting the single variable that creates the task. The task that got overwritten doesn't have a handle referencing it anymore, therefore when it enters one of the sync objects later in the task, it could be cleaned up by the GC and that will corrupt the sync object possibly not allowing anything else in. My PR a few months ago was to get rid of all the tasks entirely but this leaves things as is.

Reply:
Added comment